### PR TITLE
GODRIVER-1407 add MapCodec for mgocompat

### PR DESCRIPTION
**Pull request by iwysiu**
I realized when I did this that I missed a structcodec option

### CODE REVIEW

**Comment by iwysiu** [comment]
Sorry about the force push, I needed to rebase on the nil interface fix.

File: bson/bsoncodec/map_codec.go [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+	dr, err := vr.ReadDocument()
+	if err != nil {
+		return err
+	}
+
+	if val.IsNil() {
+		val.Set(reflect.MakeMap(val.Type()))
+	}
+
+	if val.Len() > 0 && mc.DecodeZerosMap {
```
Comment (divjotarora) [inline bson/bsoncodec/map_codec.go]: Is this the only behavioral change? I assume the rest of the code is taken from the old `DefaultValueEncoders`/`DefaultValueDecoders` functions.

File: bson/bsoncodec/struct_codec.go [hunk 2]
```diff
@@ -361,8 +366,8 @@ func (sc *StructCodec) describeStruct(r *Registry, t reflect.Type) (*structDescr
 
 	for i := 0; i < numFields; i++ {
 		sf := t.Field(i)
-		if sf.PkgPath != "" {
-			// unexported, ignore
+		if sf.PkgPath != "" && (!sc.AllowUnexportedFields || !sf.Anonymous) {
```
Comment (divjotarora) [inline bson/bsoncodec/struct_codec.go]: Why do we have a new check for `Anonymous`?

File: bson/bsonoptions/map_codec_options.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+type MapCodecOptions struct {
+	DecodeZerosMap *bool // Specifies if we should zero the map before decoding into it. Defaults to false.
+}
+
+// MapCodec creates a new *MapCodecOptions
+func MapCodec() *MapCodecOptions {
+	return &MapCodecOptions{}
+}
+
+// SetDecodeZerosMap specifies if we should zero the map before decoding into it. Defaults to false.
```
Comment (divjotarora) [inline bson/bsonoptions/map_codec_options.go]: nit: Instead of "we should zero", change to "the map should be zeroed"

File: bson/bsonoptions/string_codec_options.go [hunk 4]
```diff
@@ -15,7 +15,7 @@ type StringCodecOptions struct {
 
 // StringCodec creates a new *StringCodecOptions
 func StringCodec() *StringCodecOptions {
-	return &StringCodecOptions{&defaultDecodeOIDAsHex}
```
Comment (divjotarora) [inline bson/bsonoptions/string_codec_options.go]: Why this change?

File: bson/bsoncodec/map_codec.go (hunk 1 above)
Comment (iwysiu) [reply bson/bsoncodec/map_codec.go]: The changes are:
- clearing the map before decoding
- not checking CanSet() on a non-nil map
- allowing decimal keys for decoding
- allowing anything for encoding keys
